File: crates/brainwires-inference/examples/validation_loop.rs
//! Example: ValidationLoop quality gates before agent completion
//!
//! Demonstrates how `run_validation` enforces checks (file existence,
//! duplicate detection, syntax validity) on an agent's working set.
//! A file with intentional issues is created, validated (fails), fixed,
//! then re-validated (passes).
//!
//! Run: cargo run -p brainwires-agent --example validation_loop

use std::fs;

use anyhow::Result;

use brainwires_inference::validation_loop::{
    ValidationCheck, ValidationConfig, ValidationSeverity, format_validation_feedback,
    run_validation,
};

#[tokio::main]
async fn main() -> Result<()> {
    println!("=== Validation Loop Demo ===\n");

    // ── 1. Create a temp directory with a file that has issues ──────────────

    let tmp_dir = tempfile::tempdir()?;
    let tmp_path = tmp_dir.path().to_string_lossy().to_string();

    let bad_file = tmp_dir.path().join("utils.ts");
    let bad_content = r#"export function greet(name: string): string {
    return `Hello, ${name}!`;
}

export function greet(name: string): string {
    return `Hi, ${name}!`;
}
"#;
    fs::write(&bad_file, bad_content)?;
    println!(
        "Created file with duplicate function: {}",
        bad_file.display()
    );

    // ── 2. Configure validation ────────────────────────────────────────────

    let config = ValidationConfig {
        checks: vec![ValidationCheck::NoDuplicates, ValidationCheck::SyntaxValid],
        working_directory: tmp_path.clone(),
        max_retries: 3,
        enabled: true,
        working_set_files: vec!["utils.ts".to_string()],
        intended_writes: None,
    };

    println!("Validation config:");
    println!("  Checks:            {:?}", config.checks);
    println!("  Working directory:  {}", config.working_directory);
    println!("  Max retries:       {}", config.max_retries);
    println!("  Working set files: {:?}", config.working_set_files);

    // ── 3. Run validation — expect failure ─────────────────────────────────

    println!("\n--- First validation run (expecting issues) ---");

    let result = run_validation(&config).await?;
    println!("Passed: {}", result.passed);
    println!("Issues found: {}", result.issues.len());

    for issue in &result.issues {
        let severity = match issue.severity {
            ValidationSeverity::Error => "ERROR",
            ValidationSeverity::Warning => "WARN",
            ValidationSeverity::Info => "INFO",
        };
        println!("  [{}] {}: {}", severity, issue.check, issue.message);
        if let Some(file) = &issue.file {
            print!("         File: {file}");
            if let Some(line) = issue.line {
                print!(" (line {line})");
            }
            println!();
        }
    }

    // Show the formatted feedback an agent would receive
    let feedback = format_validation_feedback(&result);
    println!("\nAgent feedback:\n{feedback}");

    // ── 4. Fix the file ────────────────────────────────────────────────────

    println!("--- Fixing the file ---");

    let good_content = r#"export function greet(name: string): string {
    return `Hello, ${name}!`;
}

export function farewell(name: string): string {
    return `Goodbye, ${name}!`;
}
"#;
    fs::write(&bad_file, good_content)?;
    println!("Replaced duplicate function with unique 'farewell'");

    // ── 5. Re-run validation — expect pass ─────────────────────────────────

    println!("\n--- Second validation run (expecting pass) ---");

    let result2 = run_validation(&config).await?;
    println!("Passed: {}", result2.passed);
    println!("Issues found: {}", result2.issues.len());

    let feedback2 = format_validation_feedback(&result2);
    println!("Agent feedback: {feedback2}");

    // ── 6. Demonstrate file-existence check ────────────────────────────────

    println!("--- File existence check ---");

    let missing_config = ValidationConfig {
        checks: vec![],
        working_directory: tmp_path.clone(),
        max_retries: 1,
        enabled: true,
        working_set_files: vec!["nonexistent.rs".to_string()],
        intended_writes: None,
    };

    let missing_result = run_validation(&missing_config).await?;
    println!("Passed (with missing file): {}", missing_result.passed);
    for issue in &missing_result.issues {
        println!("  [{}] {}", issue.check, issue.message);
    }

    // ── 7. Summary ─────────────────────────────────────────────────────────

    println!("\n--- Summary ---");
    println!("The validation loop prevents agents from reporting success when:");
    println!("  - Files in the working set do not exist on disk (Bug #5)");
    println!("  - Duplicate exports/functions/types are present");
    println!("  - Basic syntax errors are detected");
    println!("  - Build commands fail (when BuildSuccess check is enabled)");

    println!("\nValidation loop demo complete.");
    Ok(())
}

File: deno/packages/inference/validation_loop.ts
/**
 * Validation Loop - Enforces quality checks before agent completion.
 *
 * Wraps task agent execution to automatically validate work before allowing
 * completion. If validation fails, forces the agent to fix issues.
 *
 * @module
 */

// ---------------------------------------------------------------------------
// Validation check
// ---------------------------------------------------------------------------

/** Validation checks to enforce. */
export type ValidationCheck =
  | { kind: "no_duplicates" }
  | { kind: "build_success"; buildType: string }
  | { kind: "syntax_valid" }
  | { kind: "custom_command"; command: string; args: string[] };

// ---------------------------------------------------------------------------
// Validation result
// ---------------------------------------------------------------------------

/** Result of validation checks. */
export interface ValidationResult {
  /** Whether all checks passed. */
  passed: boolean;
  /** Issues found during validation. */
  issues: ValidationIssue[];
}

/** A single issue found during validation. */
export interface ValidationIssue {
  /** Name of the check that found this issue. */
  check: string;
  /** Severity of the issue. */
  severity: ValidationSeverity;
  /** Human-readable description. */
  message: string;
  /** File where the issue was found. */
  file?: string;
  /** Line number of the issue. */
  line?: number;
}

/** Severity level for a validation issue. */
export type ValidationSeverity = "error" | "warning" | "info";

// ---------------------------------------------------------------------------
// Validation config
// ---------------------------------------------------------------------------

/** Configuration for validation loop. */
export interface ValidationConfig {
  /** Checks to run. */
  checks: ValidationCheck[];
  /** Working directory for validation. */
  workingDirectory: string;
  /** Maximum validation retry attempts. Default: 3. */
  maxRetries: number;
  /** Whether to run validation (can disable for testing). Default: true. */
  enabled: boolean;
  /** Specific files to validate (from working set). */
  workingSetFiles: string[];
}

/** Create a default ValidationConfig. */
export function defaultValidationConfig(): ValidationConfig {
  return {
    checks: [{ kind: "no_duplicates" }, { kind: "syntax_valid" }],
    workingDirectory: ".",
    maxRetries: 3,
    enabled: true,
    workingSetFiles: [],
  };
}

/** Create a disabled ValidationConfig (for testing). */
export function disabledValidationConfig(): ValidationConfig {
  return { ...defaultValidationConfig(), enabled: false };
}

// ---------------------------------------------------------------------------
// Run validation
// ---------------------------------------------------------------------------

/** Check if file is a source code file worth validating. */
function _isSourceFile(path: string): boolean {
  const lower = path.toLowerCase();
  return (
    lower.endsWith(".rs") ||
    lower.endsWith(".ts") ||
    lower.endsWith(".tsx") ||
    lower.endsWith(".js") ||
    lower.endsWith(".jsx") ||
    lower.endsWith(".py") ||
    lower.endsWith(".java") ||
    lower.endsWith(".cpp") ||
    lower.endsWith(".c") ||
    lower.endsWith(".go") ||
    lower.endsWith(".rb")
  );
}

/**
 * Run validation checks on changed files.
 *
 * Only file-existence checks and `custom_command` checks run here; a
 * `no_duplicates` or `syntax_valid` check yields a warning issue saying it is
 * not implemented (it used to pass silently). `passed` reflects errors only.
 */
export async function runValidation(
  config: ValidationConfig,
): Promise<ValidationResult> {
  if (!config.enabled) {
    return { passed: true, issues: [] };
  }
  const issues: ValidationIssue[] = [
    ...(await missingFileIssues(config)),
  ];
  for (const check of config.checks) {
    issues.push(...(await runCheck(check, config)));
  }
  return { passed: !issues.some((i) => i.severity === "error"), issues };
}

/** An error issue for every working-set file that is not on disk. */
async function missingFileIssues(
  config: ValidationConfig,
): Promise<ValidationIssue[]> {
  const issues: ValidationIssue[] = [];
  for (const file of config.workingSetFiles) {
    const filePath = config.workingDirectory === "."
      ? file
      : `${config.workingDirectory}/${file}`;
    try {
      await Deno.stat(filePath);
    } catch {
      issues.push({
        check: "file_existence",
        severity: "error",
        message:
          `File '${file}' is in working set but does not exist on disk. Agent must create file before completing.`,
        file,
      });
    }
  }
  return issues;
}

/** Run one configured check; unsupported kinds yield a warning, never a silent pass. */
async function runCheck(
  check: ValidationCheck,
  config: ValidationConfig,
): Promise<ValidationIssue[]> {
  if (check.kind === "custom_command") {
    return await runCustomCommand(
      check.command,
      check.args,
      config.workingDirectory,
    );
  }
  if (check.kind === "no_duplicates" || check.kind === "syntax_valid") {
    return [{
      check: check.kind,
      severity: "warning",
      message:
        `validation check '${check.kind}' is not implemented in the Deno port; configure a custom_command check instead`,
    }];
  }
  return [];
}

/** Run a shell-free command; a non-zero exit or spawn failure is an error issue. */
async function runCustomCommand(
  command: string,
  args: string[],
  cwd: string,
): Promise<ValidationIssue[]> {
  try {
    const output = await new Deno.Command(command, {
      args,
      cwd,
      stdout: "piped",
      stderr: "piped",
    }).output();
    if (output.success) return [];
    const stderr = new TextDecoder().decode(output.stderr);
    return [{
      check: "custom_command",
      severity: "error",
      message: `Command '${command}' failed: ${stderr}`,
    }];
  } catch (e) {
    return [{
      check: "custom_command",
      severity: "error",
      message: `Failed to run command '${command}': ${e}`,
    }];
  }
}

// ---------------------------------------------------------------------------
// Format feedback
// ---------------------------------------------------------------------------

/** Format validation result as feedback for the agent. */
export function formatValidationFeedback(result: ValidationResult): string {
  if (result.passed) {
    return "All validation checks passed!";
  }

  let feedback = "VALIDATION FAILED - You must fix these issues:\n\n";

  for (let idx = 0; idx < result.issues.length; idx++) {
    const issue = result.issues[idx];
    feedback += `${idx + 1}. [${issue.check}] `;
    if (issue.file) {
      feedback += `${issue.file}:`;
      if (issue.line != null) feedback += `${issue.line}:`;
      feedback += " ";
    }
    feedback += `${issue.message}\n`;
  }

  feedback += "\n";
  feedback +=
    "IMPORTANT: You MUST fix ALL of these issues before the task can complete.\n";
  feedback += "After fixing, verify your changes by reading the files back.\n";

  return feedback;
}
